addons/addon-base-raas/packages/base-raas-services/lib/environment/environment-mount-service.js

```diff
@@ -79,34 +80,97 @@ class EnvironmentMountService extends Service {
     };
   }
 
-  async addRoleArnToLocalResourcePolicies(workspaceRoleArn, s3Prefixes) {
+  async addRoleArnToLocalResourcePolicies(requestContext, envId, workspaceRoleArn, s3Prefixes) {
+    const envResourceUsageTrackerService = await this.service('envResourceUsageTrackerService');
     // Define function to handle updating resource policy principals where the current principals
     // may be an array or a string
-    const updateAwsPrincipals = (awsPrincipals, newPrincipal) => {
-      if (Array.isArray(awsPrincipals)) {
-        awsPrincipals.push(newPrincipal);
-      } else {
-        awsPrincipals = [awsPrincipals, newPrincipal];
+    // This function uses addEnvironmentResourceUse to track use of this resource by this principal.
+    // If this is the first use by these principals add them to the principals array.
+    const maybeUpdateAwsPrincipals = async (awsPrincipals, newPrincipal, resource) => {
+      const result = await envResourceUsageTrackerService.addEnvironmentResourceUse(requestContext, {
+        resource,
+        principal: newPrincipal,
+        envId,
+      });
+      // If this is the first environment that requires sharing the resource with the principal
+      // then update
+      if (result.count === 1) {
+        if (Array.isArray(awsPrincipals)) {
+          awsPrincipals.push(newPrincipal);
+        } else {
+          awsPrincipals = [awsPrincipals, newPrincipal];
```

Comment:
This might be problematic, because you are changing the reference of a passed parameter.